### PR TITLE
Owls 105610 - Provide ability to specify the resources and java options for the introspector pod

### PR DESCRIPTION
**Pull request by ankedia**
Owls 105610 - Provide ability to specify the resources and java options for the introspector pod. The resources and env variables for the introspector job pod can be configured by setting `domain.spec.introspector.serverPod.resources` and `domain.spec.introspector.serverPod.env` respectively. The documentation will be updated later probably in a new PR.

Integration test results - https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/16096/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/16095/

### CODE REVIEW

**Comment by tbarnes-us** [comment]
Two comments:
- Up to this point, we've been careful not to give too much flexibility in setting env vars because they are complex and easy to get wrong, and as its important for the introspector to closely reflect the same env vars that are in use by the domain pods. Is there a specific reason why the introspector pod needs to customize env vars? Could we instead add support for specific new env vars which (carefully) extend/overwrite current env vars with a carefully described precedence (similar to the way we handle tuning the node manager and WLS).

**Comment by tbarnes-us** [comment]
Two comments:
- Up to this point, we've been careful not to give too much flexibility in setting env vars because they are complex and easy to get wrong, and as its important for the introspector to closely match the env vars that are in use by the domain pods. Is there a specific reason why the introspector pod needs to customize env vars? Could we instead add support for specific new env vars which (carefully) extend/overwrite current env vars with a carefully described precedence (similar to the way we handle tuning the node manager and WLS).
- The tuning and env var documentation should be updated to reflect the new feature.

File: documentation/domains/Cluster.md [hunk 1]
```diff
@@ -23,7 +23,7 @@ The specification of the operation of the WebLogic cluster. Required.
...
 | `restartVersion` | string | Changes to this field cause the operator to restart WebLogic Server instances. More info: https://oracle.github.io/weblogic-kubernetes-operator/userguide/managing-domains/domain-lifecycle/startup/#restarting-servers. |
-| `serverPod` | [Server Pod](#server-pod) | Customization affecting the generation of Pods for WebLogic Server instances. |
```
Comment (tbarnes-us) [inline documentation/domains/Cluster.md:26]: I don't think that cluster level 'serverPod' tuning has an effect a domain level introspector. The wording could be "Customization affecting the generation of Pods for the cluster's WebLogic Server instances."

File: documentation/domains/Domain.md [hunk 2]
```diff
@@ -43,7 +44,7 @@ The specification of the operation of the WebLogic domain. Required.
...
 | `restartVersion` | string | Changes to this field cause the operator to restart WebLogic Server instances. More info: https://oracle.github.io/weblogic-kubernetes-operator/userguide/managing-domains/domain-lifecycle/startup/#restarting-servers. |
-| `serverPod` | [Server Pod](#server-pod) | Customization affecting the generation of Pods for WebLogic Server instances. |
+| `serverPod` | [Server Pod](#server-pod) | Customization affecting the generation of Pods for WebLogic Server instances or Introspector Job Pod. |
```
Comment (tbarnes-us) [inline documentation/domains/Domain.md:47]: 'the Introspector Job Pod'

File: documentation/domains/Cluster.json [hunk 3]
```diff
@@ -11,7 +11,7 @@
           "$ref": "#/definitions/ServerService"
         },
         "serverPod": {
-          "description": "Customization affecting the generation of Pods for WebLogic Server instances.",
+          "description": "Customization affecting the generation of Pods for WebLogic Server instances or Introspector Job Pod.",
```
Comment (tbarnes-us) [inline documentation/domains/Cluster.json:14]: Cluster level serverPod settings are not expected to have any effect on the introspector pod.

File: documentation/domains/Cluster.json [hunk 4]
```diff
@@ -86,6 +86,22 @@
         }
       }
     },
+    "BaseServerPod": {
```
Comment (tbarnes-us) [inline documentation/domains/Cluster.json:89]: It doesn't look like this belongs in the cluster scope. Also, the name 'BaseServerPod' looks internal/unintended.

File: documentation/domains/Cluster.json [hunk 5]
```diff
@@ -509,7 +529,7 @@
           "type": "boolean"
         },
         "serverPod": {
-          "description": "Customization affecting the generation of Pods for WebLogic Server instances.",
+          "description": "Customization affecting the generation of Pods for WebLogic Server instances or Introspector Job Pod.",
```
Comment (tbarnes-us) [inline documentation/domains/Cluster.json:532]: "Introspector Job Pod' doesn't look like it belongs in the cluster scope.

File: documentation/domains/Cluster.json [hunk 6]
```diff
@@ -645,6 +665,15 @@
         }
       }
     },
+    "Introspector": {
```
Comment (tbarnes-us) [inline documentation/domains/Cluster.json:668]: This stanza doesn't look like it belongs at the cluster scope.

File: documentation/domains/Domain.json [hunk 7]
```diff
@@ -86,6 +86,22 @@
         }
       }
     },
+    "BaseServerPod": {
```
Comment (tbarnes-us) [inline documentation/domains/Domain.json:89]: This looks like an internal artifact that doesn't belong in the public schema?

File: documentation/domains/Domain.json [hunk 8]
```diff
@@ -11,7 +11,7 @@
           "$ref": "#/definitions/ServerService"
         },
         "serverPod": {
-          "description": "Customization affecting the generation of Pods for WebLogic Server instances.",
+          "description": "Customization affecting the generation of Pods for WebLogic Server instances or Introspector Job Pod.",
```
Comment (tbarnes-us) [inline documentation/domains/Domain.json:14]: 'the Introspector Job Pod'

File: documentation/domains/Domain.json [hunk 9]
```diff
@@ -576,7 +605,7 @@
           "type": "string"
         },
         "serverPod": {
-          "description": "Customization affecting the generation of Pods for WebLogic Server instances.",
+          "description": "Customization affecting the generation of Pods for WebLogic Server instances or Introspector Job Pod.",
```
Comment (tbarnes-us) [inline documentation/domains/Domain.json:608]: I don't think the 'ManagedServer' scope can have an affect on the introspector pod, which I think only historically pays attention to 'adminServer' and 'domain' level scopes.

**Review by tbarnes-us — COMMENTED** [review]
I have added multiple comments related to scoping.

Assuming I'm correct, then the unit tests should verify introspector scoped changes such that:

- introspector scoped changes have no effect on wl instances
- the introspector gets a merged view of the env with precedence from highest to lowest of 'introspector.serverPod', 'adminServer.serverPod', 'domain.serverPod' 
- managed server or cluster scoped serverPod changes have no effect on the generated introspector env

File: documentation/domains/Cluster.md (hunk 1 above)
Comment (ankedia) [reply documentation/domains/Cluster.md:26]: The `serverPod` is shared between domain, cluster and server scopes. I can't give different documentation for different scopes. I'll start a slack thread to discuss this.

File: common/src/main/resources/Operator.properties [hunk 10]
```diff
@@ -325,4 +325,7 @@ WLSWH-0020=Change request to domain resource ''{0}'' cannot be honored because t
 WLSWH-0021=Change request to domain resource ''{0}'' causes the replica count of each cluster in ''{1}'' to exceed its cluster \
   size ''{2}'' respectively
 WLSWH-0022=Scale request to cluster resource ''{0}'' cannot be honored because the replica count would exceed the \
-  cluster size ''{1}''
\ No newline at end of file
+  cluster size ''{1}''
+WLSWH-0023=Unsupported introspector environment variable(s) ''{0}'' defined in ''introspector.serverPod.env'', \
+  supported values are ''JAVA_OPTIONS'', ''USER_MEM_ARGS'', ''NODEMGR_JAVA_OPTIONS'' and ''NODEMGR_MEM_ARGS''
```
Comment (tbarnes-us) [inline common/src/main/resources/Operator.properties:330]: Minor: Consider passing in the list as I assume this is coded up in a central place in Java.

File: documentation/domains/Cluster.json (hunk 6 above)
Comment (ankedia) [reply documentation/domains/Cluster.json:652]: There is a general issue where all domain level stanza are included at the cluster scope incuding 'AdminServer` and `domainHomeSourceType` etc. This needs to be investigated seprately.

File: documentation/domains/Domain.md (hunk 2 above)
Comment (ankedia) [reply documentation/domains/Domain.md]: Changed.

File: documentation/domains/Cluster.json (hunk 3 above)
Comment (ankedia) [reply documentation/domains/Cluster.json]: There's a general issue with Cluster.json where all domain level settings are included in Cluster.json. This needs to be investigated separately.

File: documentation/domains/Cluster.json (hunk 4 above)
Comment (ankedia) [reply documentation/domains/Cluster.json]: Renamed to `IntrospectorPod`.

File: documentation/domains/Cluster.json (hunk 5 above)
Comment (ankedia) [reply documentation/domains/Cluster.json]: There's a general issue with Cluster.json where all domain level settings are included in Cluster.json. This needs to be investigated separately.

File: documentation/domains/Domain.json (hunk 7 above)
Comment (ankedia) [reply documentation/domains/Domain.json]: Renamed to `IntrospectorPod`.

File: documentation/domains/Domain.json (hunk 8 above)
Comment (ankedia) [reply documentation/domains/Domain.json]: Changed.

File: documentation/domains/Domain.json (hunk 9 above)
Comment (ankedia) [reply documentation/domains/Domain.json]: The serverPod is shared between domain, cluster and server scopes. Its currently not possible to give different description at different scopes.

File: common/src/main/resources/Operator.properties (hunk 10 above)
Comment (ankedia) [reply common/src/main/resources/Operator.properties]: Changed.

File: documentation/domains/Cluster.json (hunk 3 above)
Comment (ankedia) [reply documentation/domains/Cluster.json]: Fixed in [a6b2383](https://github.com/oracle/weblogic-kubernetes-operator/pull/3914/commits/a6b23837e674e333f027e573edba8c9dedd57ca5).

File: documentation/domains/Cluster.json (hunk 5 above)
Comment (ankedia) [reply documentation/domains/Cluster.json]: Fixed in [a6b2383](https://github.com/oracle/weblogic-kubernetes-operator/pull/3914/commits/a6b23837e674e333f027e573edba8c9dedd57ca5).

File: documentation/domains/Cluster.json (hunk 6 above)
Comment (ankedia) [reply documentation/domains/Cluster.json]: Fixed in [a6b2383](https://github.com/oracle/weblogic-kubernetes-operator/pull/3914/commits/a6b23837e674e333f027e573edba8c9dedd57ca5).

**Review by rjeberhard — APPROVED** [review]

**Review by jshum2479 — APPROVED** [review]

**Pull request merged**